File: app/backend/src/application/queries/handlers/get_product_list_handler.rs
use std::sync::Arc;

use crate::application::repositories::ProductRepository;
use crate::application::error::ApplicationError;
use crate::application::dto::ProductListDTO;

/// 商品リスト取得クエリハンドラ
/// CQRS パターンに基づく読み取り操作のハンドラ
pub struct GetProductListHandler {
    product_repository: Arc<dyn ProductRepository + Send + Sync>,
}

impl GetProductListHandler {
    pub fn new(product_repository: Arc<dyn ProductRepository + Send + Sync>) -> Self {
        Self {
            product_repository,
        }
    }

    /// 商品リスト取得クエリを実行
    /// 
    /// # Returns
    /// * `Result<ProductListDTO, ApplicationError>` - 成功時は商品リストデータ、失敗時はエラー
    pub async fn handle(&self) -> Result<ProductListDTO, ApplicationError> {
        println!("->> get_product_list_handler");
        
        let product_list = self.product_repository.find_all().await?;

        Ok(product_list)
    }
} 

File: app/backend-ts/src/application/queries/handlers/get-product-list.handler.ts
import { Inject } from '@nestjs/common';
import { IQueryHandler, QueryHandler } from '@nestjs/cqrs';

import { ProductListDto } from '$application/dto';
import {
  NotFoundError,
  ValidationError,
} from '$application/errors/application.error';
import { GetProductListQuery } from '$application/queries';
import {
  ICategoryRepository,
  IProductRepository,
} from '$application/repositories';
import { CategoryId } from '$domain/value-objects';

@QueryHandler(GetProductListQuery)
export class GetProductListHandler
  implements IQueryHandler<GetProductListQuery>
{
  constructor(
    @Inject('IProductRepository')
    private readonly productRepository: IProductRepository,
    @Inject('ICategoryRepository')
    private readonly categoryRepository: ICategoryRepository,
  ) {}

  async execute(query: GetProductListQuery): Promise<ProductListDto> {
    // Validate and parse category filter
    let categoryId: CategoryId | undefined;
    if (query.categoryId) {
      try {
        categoryId = CategoryId.fromUuid(query.categoryId);

        // Verify category exists
        const category = await this.categoryRepository.findById(categoryId);
        if (!category) {
          throw new NotFoundError('Category', query.categoryId);
        }
      } catch (error: unknown) {
        if (
          error instanceof ValidationError ||
          error instanceof NotFoundError
        ) {
          throw error;
        }
        throw new ValidationError('Invalid category ID format');
      }
    }

    // Fetch products with pagination handled by repository
    const productListDto = categoryId
      ? await this.productRepository.findByCategory(categoryId)
      : await this.productRepository.findAll();

    return productListDto;
  }
}
